relay-server/src/extractors/decoder.rs

```diff
@@ -0,0 +1,212 @@
+        if !buf.is_empty() && self.buffer.is_empty() {
+            self.buffer.reserve(DECODE_BUFFER_SIZE.min(self.remaining));
+        }
+
+        self.buffer.extend_from_slice(buf);
+        self.remaining -= buf.len();
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        Ok(())
+    }
+}
+
+/// Writes data into the given `write`, returning `true` on overflow.
+fn write_overflowing<W: Write>(write: &mut W, slice: &[u8]) -> io::Result<bool> {
+    match write.write_all(slice).and_then(|()| write.flush()) {
+        Ok(()) => Ok(false),
+        Err(e) => match e.kind() {
+            io::ErrorKind::WriteZero => Ok(true),
+            _ => Err(e),
+        },
+    }
+}
+
+/// Internal dispatch for all supported [`HttpEncoding`]s.
+enum DecoderInner {
+    Identity(Box<Sink>),
+    Br(Box<BrotliDecoder<Sink>>),
+    Gzip(Box<GzDecoder<Sink>>),
+    Deflate(Box<DeflateDecoder<Sink>>),
+}
+
+/// Stateful decoder for all supported [`HttpEncoding`]s.
+///
+/// Use [`decode`](Self::decode) to feed data into the decoder's internal buffer. To read the
+/// intermediate buffer, use [`take`](Self::take). Decoding can continue afterwards, but the taken
+/// payload will not be returned again.
+///
+/// To initialize the decoder, assign a `limit`. When the decoded buffer reaches the size limit, it
+/// is truncated and an overflow is returned.
+pub struct Decoder {
+    inner: DecoderInner,
+}
+
+impl Decoder {
+    /// Creates a new `Decoder` with the given size limit.
+    ///
+    /// If the request is not encoded, this decoder is a noop.
+    pub fn new<S>(request: &HttpRequest<S>, limit: usize) -> Self {
+        let sink = Sink::new(limit);
+
+        let encoding = request
+            .headers()
+            .get(CONTENT_ENCODING)
+            .and_then(|enc| enc.to_str().ok())
+            .map(HttpEncoding::parse)
+            .unwrap_or_default();
+
+        let inner = match encoding {
+            HttpEncoding::Identity => DecoderInner::Identity(Box::new(sink)),
+            HttpEncoding::Br => DecoderInner::Br(Box::new(BrotliDecoder::new(sink))),
+            HttpEncoding::Gzip => DecoderInner::Gzip(Box::new(GzDecoder::new(sink))),
+            HttpEncoding::Deflate => DecoderInner::Deflate(Box::new(DeflateDecoder::new(sink))),
+        };
+
+        Self { inner }
+    }
+
+    /// Decodes a chunk of data.
+    ///
+    /// The decoded bytes are written into the Decoder's buffer, which can be obtained using
+    /// [`take`](Self::take). Returns `Ok(false)` if decoding has completed successfully without an
+    /// overflow. Returns `Ok(true)` if decoding has stopped prematurely due to an overflow. In this
+    /// case, the buffer contains the decoded payload up to the limit. Returns `Err` if there was an
+    /// error decoding.
+    pub fn decode(&mut self, bytes: Bytes) -> io::Result<bool> {
+        match &mut self.inner {
+            DecoderInner::Identity(inner) => write_overflowing(inner, &bytes),
```

Comment:
In theory, this causes one unnecessary allocation for bodies that can be fully streamed in a single chunk. To optimize this, we could create a bespoke implementation here that retains the original bytes and only joins them in `take()`. 

In practice, I do not expect such an optimization to have a large effect, so I went with the simpler uniform implementation.